### PR TITLE
TS: raise the Node support floor to >=22.12.0, simplify to AbortSignal.any (closes #421)

### LINKED ISSUE #421: TS: brace-expansion@5 requires node >=20 while package.json engines say >=18
Follow-up from a resolved Copilot thread on #412 ([discussion](https://github.com/basecamp/basecamp-sdk/pull/412#discussion_r3649296058)), filed for durable tracking.

### Observation
`typescript/package-lock.json` now pulls `brace-expansion@5` (transitive dev-tooling dep), which declares `engines.node: "20 || >=22"`, while `typescript/package.json` still advertises:

```json
"engines": { "node": ">=18.0.0" }
```

### Impact (non-blocking)
There is no `.npmrc` and `engine-strict` is unset, so npm treats the mismatch as a **warning**, never an `npm install`/`npm ci` failure — CI is unaffected today. It is a stated-support-range inconsistency, not a break.

### Decision needed
A Node-18-support policy call: either
- (a) drop Node 18 and raise `engines.node` to `>=20` (aligning with the transitive floor and dropping the Node 18 mentions in the runtime code), or
- (b) keep advertising `>=18` and accept that a dev-tooling dep floors at 20 (harmless without engine-strict).

No urgency; capturing so the range stays intentional rather than drifting.

### PR DESCRIPTION
**Pull request by jeremy**
Closes #421, taking option **(a)** — drop Node 18, and drop the Node-18 mentions in the runtime code.

## Why `>=22.12.0` rather than the issue's `>=20`

`typescript/package.json` advertised `>=18.0.0`. That was an **untested claim**, and the issue's suggested `>=20` would replace it with a **dead** one:

| | |
|---|---|
| Node 18 | EOL **2025-04-30** |
| Node 20 | EOL **2026-04-30** |
| CI (`test`, `security`, `live-canary`, `release-typescript` build) | `node-version: '22'` |
| CI (`release-typescript` publish) | `node-version: '24'` |
| `.mise.toml` | `node = "22"` |
| `conformance/runner/typescript/package.json` | `^20.19.0 \|\| >=22.12.0` |

Nothing exercises 18 or 20. `>=22.0.0` would be no better — a floating `node-version: '22'` resolves to the latest 22.x, so 22.0 is never tested either.

`>=22.12.0` is the first Node 22 LTS **and** the floor the conformance runner already declares, so the advertised range now lines up with a version we actually run.

**Residual honesty gap:** floating `'22'` currently resolves to 22.23.1, not 22.12.0, so the exact floor is still not pinned by a job. If we want that guarantee, the right fix is a minimum-version CI entry — deliberately not bundled here.

## `AbortSignal.any`

The floor makes `AbortSignal.any` (Node ≥ 20.3) available, replacing the hand-rolled plumbing:

```diff
-      // Apply request timeout (Node 18-compatible: no AbortSignal.any)
-      const controller = new AbortController();
-      setTimeout(() => controller.abort(), requestTimeoutMs);
-      if (request.signal) {
-        request.signal.addEventListener("abort", () => controller.abort(), { once: true });
-      }
+      const timeoutSignal = AbortSignal.timeout(requestTimeoutMs);
+      const signal = request.signal
+        ? AbortSignal.any([request.signal, timeoutSignal])
+        : timeoutSignal;
```

**This also fixes a timer leak.** The old `setTimeout` was never cleared, so every request left a live timer behind until it fired. `AbortSignal.timeout` uses an unref'd timer.

**Behavior change worth knowing:** a timed-out request now rejects with `TimeoutError` instead of a generic `AbortError`, which distinguishes a timeout from a caller cancellation. Nothing in the SDK classifies abort errors by name (checked `errors.ts`), but a consumer matching on `err.name === "AbortError"` for timeouts would need to update — hence the `breaking` label, alongside the engines raise itself.

## Tests

Two tests, one per input of the combined signal, because a regression could silently drop either:

1. **timeout fires with no caller signal** — 2s handler, 50ms timeout, asserts it aborts in <1s rather than waiting the handler out.
2. **caller signal propagates** — 30s timeout so it cannot be what aborts, caller aborts at 50ms, asserts prompt rejection.

Test 2 was **verified to fail** when the `AbortSignal.any` composition is removed, so it is not vacuous.

## Verification

`make ts-check` — 68 files, **795 tests**, all green; `tsc --noEmit` clean; `ts-check-drift` clean. `npm ci --engine-strict` passes on the declared floor (running 22.23.1).

---
## Summary by cubic
Raises the Node engine floor to >=22.12.0 and switches to AbortSignal.any/AbortSignal.timeout to fix the timer leak and return TimeoutError on timeouts. Validates `requestTimeoutMs` as an integer between 0 and 2,147,483,647. Closes #421.

- **Migration**
  - Requires Node >=22.12.0 across the repo (`README`, `CONTRIBUTING.md`, lockfiles, runner and root `package.json`).
  - Timeouts now reject with TimeoutError; `requestTimeoutMs` must be an integer in [0..2,147,483,647].

- **Refactors**
  - Use AbortSignal.any with AbortSignal.timeout; the timeout’s timer is unref’d.
  - Tests assert error identities, widen timing cutoff, add table-driven validation of `requestTimeoutMs`, and use catch + toMatchObject for clearer failure output.

<sup>Written for commit d715a23c7fc98a06601d9211b5d3a1c0cc5d61ef. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/basecamp/basecamp-sdk/pull/469?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: typescript/package.json [hunk 1]
```diff
@@ -47,7 +47,7 @@
     "directory": "typescript"
   },
   "engines": {
-    "node": ">=18.0.0"
+    "node": ">=22.12.0"
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Addressed at exact head 2a5d2c1: typescript/README.md now states Node.js 22.12+, matching package.json.

File: typescript/package.json [hunk 2]
```diff
@@ -47,7 +47,7 @@
...
   },
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:51]: Addressed at exact head 2a5d2c1: typescript/README.md now states Node.js 22.12+, matching package.json.

File: typescript/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Addressed at exact head 2a5d2c1: typescript/README.md now states Node.js 22.12+, matching package.json.

[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Addressed at exact head 2a5d2c1: the package-lock root engines entry is regenerated to >=22.12.0.

[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Fixed — `typescript/README.md:26` now reads "Requires Node.js 22.12+". Confirmed npm packs the README regardless of the `files` allowlist, so the published package was advertising a floor on which every request throws `AbortSignal.any is not a function`.

File: typescript/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:51]: Fixed — `typescript/README.md:26` updated to "Requires Node.js 22.12+ and TypeScript 5.0+." It was the only user-facing doc carrying the old floor.

File: typescript/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Fixed — README updated to 22.12+.

[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Fixed — regenerated with `npm install --package-lock-only`; a one-line engines change, no dependency churn. Also caught a second stale entry the reviews didn't flag: `conformance/runner/typescript/package-lock.json` recorded `>=18.0.0` on its `../../../typescript` link entry. Both are updated.

File: typescript/tests/client.test.ts [hunk 3]
```diff
@@ -416,6 +416,63 @@ describe("BasecampClient", () => {
...
+
+      const client = createBasecampClient({
+        accountId: "12345",
+        accessToken: "test-token",
+        enableRetry: false,
+        requestTimeoutMs: 50,
+      });
+
+      const startedAt = Date.now();
+      await expect(client.GET("/projects.json")).rejects.toThrow();
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/tests/client.test.ts]: Fixed — the timeout test now asserts `name: "TimeoutError"` and the caller-abort test asserts `name: "AbortError"`. Verified non-vacuous: restoring the old AbortController+setTimeout timeout makes the TimeoutError assertion fail with AbortError.

[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/tests/client.test.ts]: Fixed — timeout path now asserts `name: "TimeoutError"`, locking in the new contract. Verified it fails against the previous AbortError behavior.

File: typescript/tests/client.test.ts [hunk 4]
```diff
@@ -416,6 +416,63 @@ describe("BasecampClient", () => {
...
+        requestTimeoutMs: 30000,
+      });
+
+      const controller = new AbortController();
+      setTimeout(() => controller.abort(), 50);
+
+      const startedAt = Date.now();
+      await expect(
+        client.GET("/projects.json", { signal: controller.signal })
+      ).rejects.toThrow();
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/tests/client.test.ts]: Fixed — asserts `name: "AbortError"`, which also proves the caller's signal won rather than the 30s timeout. The abort timer is now cleared in a `finally`.

File: typescript/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Fixed — root `package.json` raised to >=22.12.0 (its `main`/`exports` point at `typescript/dist/index.js`, so installing from the repo root really does get this SDK), and `CONTRIBUTING.md:12` now reads "Node.js 22.12+, npm".

File: typescript/src/client.ts [hunk 5]
```diff
@@ -442,20 +442,19 @@ function createAuthMiddleware(authStrategy: AuthStrategy, userAgent: string, req
...
-          once: true,
-        });
-      }
+      // Apply request timeout, preserving any caller-supplied signal.
+      // AbortSignal.timeout's timer is unref'd, so unlike a bare setTimeout it
+      // never holds the event loop open after the request settles.
+      const timeoutSignal = AbortSignal.timeout(requestTimeoutMs);
+      const signal = request.signal
+        ? AbortSignal.any([request.signal, timeoutSignal])
+        : timeoutSignal;
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts:477]: Fixed — `requestTimeoutMs` is now validated at construction and throws `BasecampError("usage", ...)`, matching the existing auth/baseUrl config checks. You are right that it was a behavior change: the old `setTimeout` coerced NaN/negative to 0, whereas `AbortSignal.timeout` throws a bare `RangeError` per request, far from the call that misconfigured it. Zero is still allowed (immediate abort).

File: typescript/tests/client.test.ts [hunk 6]
```diff
@@ -416,6 +416,75 @@ describe("BasecampClient", () => {
...
+    // floor). Both inputs have to stay live: the timeout must still fire when
+    // the caller passes no signal, and a caller abort must still win when it
+    // fires first. Test both directions so a regression in either input of the
+    // combined signal is caught.
+    it("aborts on timeout when the caller supplies no signal", async () => {
+      server.use(
+        http.get(`${BASE_URL}/projects.json`, async () => {
+          await new Promise(resolve => setTimeout(resolve, 2000));
+          return HttpResponse.json([]);
+        })
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/tests/client.test.ts:432]: Fixed — handler delays halved to 1000ms. I also tightened the assertion threshold from 1000ms to 500ms in the same edit, because simply shortening the delay would have made the tests vacuous: with a 500ms delay and a <1000ms assertion, a regressed timeout would complete inside the threshold and still pass. Verified by neutering the timeout — both tests fail as they should.

File: conformance/runner/typescript/package-lock.json [hunk 7]
```diff
@@ -36,7 +36,7 @@
         "vitest": "^4.0.18"
       },
       "engines": {
-        "node": ">=18.0.0"
+        "node": ">=22.12.0"
       },
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/typescript/package-lock.json:40]: Fixed — narrowed the runner to `>=22.12.0` and regenerated both lockfiles. Agreed the `^20.19.0` branch is no longer satisfiable now that it depends on an SDK requiring >=22.12.0.

File: typescript/src/client.ts [hunk 8]
```diff
@@ -269,6 +269,17 @@ export function createBasecampClient(options: BasecampClientOptions): BasecampCl
     throw new BasecampError("usage", "Either 'auth' or 'accessToken' is required");
   }
 
+  // AbortSignal.timeout throws a bare RangeError for a negative, NaN, or
+  // non-finite delay — per request, and far from the call that misconfigured it.
+  // (The previous setTimeout coerced such values to 0, aborting immediately.)
+  // Fail fast at construction instead, like the other config checks.
+  if (!Number.isFinite(requestTimeoutMs) || requestTimeoutMs < 0) {
```
Comment (jeremy) [inline typescript/src/client.ts]: P2: The new validation still accepts values that `AbortSignal.timeout` rejects (or mis-schedules). Verified on Node v22.23.1: `1.5` and `4294967296` pass this guard and then throw a bare `RangeError`; values above the signed 32-bit timer range also emit `TimeoutOverflowWarning` and are clamped to 1ms. Validate a non-negative integer within the usable timer range (practically `<= 2_147_483_647`), and add a table-driven construction test covering negative, NaN/infinite, fractional, and over-range values with the exact `BasecampError` surface. Also update the nearby test comment that still says “2000ms handler” after the handler was shortened to 1000ms.

File: typescript/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/package.json:50]: Addressed at exact head d49a84c: root package.json and CONTRIBUTING.md now both declare Node 22.12+.

File: typescript/src/client.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts:477]: Addressed in d49a84c with fail-fast BasecampError validation. A new exact-range review thread captures the remaining fractional/upper-bound cases.

File: conformance/runner/typescript/package-lock.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/typescript/package-lock.json:40]: Addressed at exact head d49a84c: the TypeScript conformance runner and its lockfile now require Node >=22.12.0.

File: typescript/src/client.ts (hunk 8 above)
Comment (jeremy) [reply typescript/src/client.ts]: Fixed — and confirmed all three cases on Node v22.23.1 before changing anything:

```
1.5        -> RangeError: ... must be an integer
4294967296 -> RangeError: ... must be >= 0 && <= 4294967295
2147483648 -> ACCEPTED, TimeoutOverflowWarning, duration silently clamped to 1
```

The clamping case is the worst of the three — it does not throw, it just aborts every request almost immediately. The guard is now a non-negative integer `<= 2_147_483_647` (`MAX_TIMEOUT_MS`, the signed 32-bit timer bound). Added table-driven construction tests covering negative / NaN / Infinity / fractional / 2^31 / 2^32 on the reject side and 0 / 30000 / 2147483647 on the accept side, asserting the exact `{ name, code, message }` surface rather than just that it throws. Stale "2000ms handler" comment corrected too.

File: typescript/src/client.ts [hunk 9]
```diff
@@ -79,6 +79,13 @@ export type { paths };
  * Raw client type from openapi-fetch.
  * Use this when you need direct access to GET/POST/PUT/DELETE methods.
  */
+/**
```
Comment (jeremy) [inline typescript/src/client.ts:78]: P3: This insertion strands the existing `RawClient` JSDoc. The “Raw client type from openapi-fetch” block is no longer immediately above `export type RawClient`; it now precedes a second JSDoc plus `MAX_TIMEOUT_MS`, so generated/editor docs for the public `RawClient` type lose that description. Move the timeout constant (and its own block) before the RawClient JSDoc, or move the RawClient block back directly above the export.

File: typescript/src/client.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/src/client.ts]: Already fixed — this reviewed the previous commit. The guard now uses `Number.isInteger` (plus a `<= 2_147_483_647` bound), so `0.5` is rejected at construction. Confirmed on Node v22.23.1 that fractional values throw a bare RangeError from `AbortSignal.timeout`, which is exactly what the fail-fast check exists to prevent.

File: typescript/src/client.ts (hunk 9 above)
Comment (jeremy) [reply typescript/src/client.ts:78]: Fixed — moved `MAX_TIMEOUT_MS` and its block above the RawClient JSDoc, so that block sits directly on `export type RawClient` again. Good catch; the insertion silently cost the public type its description.

File: typescript/tests/client.test.ts [hunk 10]
```diff
@@ -416,6 +416,75 @@ describe("BasecampClient", () => {
...
+        // A caller abort surfaces as AbortError, distinct from the timeout's
+        // TimeoutError — so this also proves it was the caller's signal, not
+        // the (30s) timeout, that won.
+        await expect(
+          client.GET("/projects.json", { signal: controller.signal })
+        ).rejects.toMatchObject({ name: "AbortError" });
+
+        // Aborting promptly proves the caller's signal reached the request; the
+        // 30s timeout signal could not have fired.
+        expect(Date.now() - startedAt).toBeLessThan(500);
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/tests/client.test.ts]: Fixed — widened to <900ms as suggested. Verified it still fails when the timeout is neutered, since the handler delay is 1000ms. Worth noting the timing assertion is secondary now: the tests also assert `name: "TimeoutError"` / `"AbortError"`, and a dropped signal makes the request succeed outright, which fails `rejects` regardless of timing.

File: typescript/tests/client.test.ts [hunk 11]
```diff
@@ -416,6 +416,121 @@ describe("BasecampClient", () => {
...
+          requestTimeoutMs: value,
+        })
+      ).toThrowError(
+        expect.objectContaining({
+          name: "BasecampError",
+          code: "usage",
+          message: expect.stringContaining("'requestTimeoutMs' must be an integer"),
+        })
+      );
+    });
```
[the remark this answers is not in the document]
Comment (jeremy) [reply typescript/tests/client.test.ts:527]: Adopted — but flagging one correction: `toThrowError(expect.objectContaining(...))` is **not** invalid under Vitest 4. I checked before changing it by deliberately altering the thrown `code` and `message`, and all six rejection cases failed, so the assertion was doing real work rather than silently passing.

Where you are right is the diagnostics. The old form reported only `expected error to match asymmetric matcher`; the catch + `toMatchObject` version names the offending field:

```
-   "code": "usage",
+   "code": "network",
```

Switched for that reason, with `toBeInstanceOf(BasecampError)` alongside it.
